project2/os.c

```diff
@@ -99,30 +98,53 @@ PID Task_Pid() {
 /**
  * The calling task terminates itself.
  */
-void Task_Terminate()
-{
+void Task_Terminate() {
     KERNEL_REQUEST_PARAMS info = {
         .request = TERMINATE
     };
 
     Kernel_Request(&info);
 }
 
-
 void Msg_Send(PID id, MTYPE t, uint16_t* v) {
+    KERNEL_REQUEST_PARAMS info = {
+        .request = MSG_SEND,
+        .msg_ptr_data = v,
+        .msg_mask = t,
+        .msg_to = id
+    };
 
+    Kernel_Request(&info);
+    v = &info.msg_data;
 }
 
 PID Msg_Recv(MASK m, uint16_t* v) {
-    return -1;
+    KERNEL_REQUEST_PARAMS info = {
+        .request = MSG_RECV,
+        .msg_mask = m
+    };
+
+    Kernel_Request(&info);
+    v = info.msg_ptr_data;
+    return info.out_pid;
 }
 
 void Msg_Rply(PID id, uint16_t r) {
+    KERNEL_REQUEST_PARAMS info = {
+        .request = MSG_RPLY,
+        .msg_to = id,
+        .msg_data = r
+    };
 
+    Kernel_Request(&info);
 }
 
 void Msg_ASend(PID id, MTYPE t, uint16_t v) {
+    KERNEL_REQUEST_PARAMS info = {
+        .request = MSG_ASEND
```

Comment:
Missing `.msg_data`?

Reply:
Oh wait, ASend isn't implemented yet anyway

Reply:
Yup, it should be `.msg_data` and `.msg_to` when it is implemented :)